### PR TITLE
WL 21453 - Add a "Dominant Hand" setting

### PR DESCRIPTION
**Pull request by cain-kilgore**
> Add a "Dominant Hand" setting:
> In Settings > Avatar, add two checkboxes (one of which has to be checked at all times): "Right Hand" and "Left Hand". It should be accessible by our scripting API as a property.
> Worklist: https://worklist.net/21453

## Test Plan
1. Launch the Interface
2. Open the Console and Type `MyAvatar.getDominantHand()`
3. Confirm that it returns `right`.
4. **In Desktop Mode**, go to Settings > Avatar
5. Confirm that you see the Left and Right Dominant Hand Setting under Avatar Tuning
6. Confirm that one Radio Box must be ticked at all times.
7. Confirm that when you select Left and Save Settings, that the Command From Step 2 returns `left`.
8. **In Tablet Mode**, launch the Tablet and  go to Menu > Settings > Avatar
9. Confirm that you see the Left and Right Dominant Hand Setting under Avatar Tuning
10. Confirm that one Radio Box must be ticked at all times.
11. Confirm that when you select Right and Save Settings, that the Command From Step 2 returns `right`.
12. Open the Console and Type `MyAvatar.setDominantHand("left")`
13. Confirm that it sets successfully, and re-running the Command from Step 2 returns `left`.
14. Check to ensure that each setting is updating in the Avatar Settings **in both Desktop & Tablet**.
15. Relaunch the game and confirm that the settings for both left and right stay put.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Can one of the admins verify this patch?

**Comment by ctrlaltdavid** [comment]
Test this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1879/HighFidelity-Beta-PR11029-a4cf27402dcf541374c1b13d3e938517b3740b75.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1879/HighFidelity-Beta-PR11029-a4cf27402dcf541374c1b13d3e938517b3740b75.exe

File: interface/src/avatar/MyAvatar.h [hunk 1]
```diff
@@ -339,6 +339,9 @@ class MyAvatar : public Avatar {
     Q_INVOKABLE bool getClearOverlayWhenMoving() const { return _clearOverlayWhenMoving; }
     Q_INVOKABLE void setClearOverlayWhenMoving(bool on) { _clearOverlayWhenMoving = on; }
 
+    Q_INVOKABLE void setUseAlternativeHand(bool hand) { _useAlternativeHand = hand; }
+    Q_INVOKABLE bool getUseAlternativeHand() const { return _useAlternativeHand; }
```
Comment (huffman) [inline interface/src/avatar/MyAvatar.h]: Is there a reason the API doesn't expose something similar to the setting itself, like `getDominantHand`, and return `'right'` or `'left'`? This seems like what was implied by the worklist item, and seems clearer as to what the return value means. Thoughts?

Comment (cain-kilgore) [inline interface/src/avatar/MyAvatar.h]: I was having a think about this - and I could potentially have it return the Controllers.Standard.LeftHand and Controllers.Standard.RightHand and I could change it up a bit. What do you think?

Comment (cain-kilgore) [inline interface/src/avatar/MyAvatar.h]: I would need to look into how to do that though. Returning true/false for 'useDominantHand' seemed easiest at the time.

File: interface/resources/qml/dialogs/preferences/PrimaryHandPreference.qml [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            width: 60
+            anchors {
+                right: parent.right
+                verticalCenter: parent.verticalCenter
+            }
+            onClicked: {
+                if(box1.checked) {
+                    box1.checked = false;
+                }
+                if(!box1.checked && !box2.checked) {
```
Comment (huffman) [inline interface/resources/qml/dialogs/preferences/PrimaryHandPreference.qml]: Style: there should be a space after `if` here and above.

**Comment by themelissabrown** [comment]
add to whitelist

File: interface/src/avatar/MyAvatar.h [hunk 3]
```diff
@@ -339,6 +339,14 @@ class MyAvatar : public Avatar {
     Q_INVOKABLE bool getClearOverlayWhenMoving() const { return _clearOverlayWhenMoving; }
     Q_INVOKABLE void setClearOverlayWhenMoving(bool on) { _clearOverlayWhenMoving = on; }
 
+    Q_INVOKABLE void setDominantHand(const QString& hand) {
+        if (hand == "left" || hand == "right") {
+            _dominantHand = hand;
+        }
+    }
```
Comment (huffman) [inline interface/src/avatar/MyAvatar.h]: This is now taking up multiple lines - can we move it to the .cpp?

File: interface/src/avatar/MyAvatar.h [hunk 4]
```diff
@@ -339,6 +339,14 @@ class MyAvatar : public Avatar {
     Q_INVOKABLE bool getClearOverlayWhenMoving() const { return _clearOverlayWhenMoving; }
     Q_INVOKABLE void setClearOverlayWhenMoving(bool on) { _clearOverlayWhenMoving = on; }
 
+    Q_INVOKABLE void setDominantHand(const QString& hand) {
+        if (hand == "left" || hand == "right") {
```
Comment (huffman) [inline interface/src/avatar/MyAvatar.h]: Can "right" and "left" be defined as const variables somewhere, and used in place of bare strings (here and elsewhere)?

**Comment by huffman** [comment]
@cain-kilgore Couple more comments, but looks good otherwise

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1882/HighFidelity-Beta-PR11029-42742ba1f9fda98db0fd11281a5b68b5a471fa09.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1882/HighFidelity-Beta-PR11029-42742ba1f9fda98db0fd11281a5b68b5a471fa09.exe

**Comment by Nex-Pro** [comment]
Testing.

[UncaughtException evaluate] Error: Result of expression 'MyAvatar.getDominantHand' [undefined] is not a function. in about:console:1 [Backtrace] <global>() at about:console:1
TypeError: Result of expression 'MyAvatar.getDominantHand' [undefined] is not a function.

Will await fresh build

**Comment by Nex-Pro** [comment]
@cain-kilgore ^

**Comment by cain-kilgore** [comment]
Weird.. thanks @Nex-Pro: Looking into it!

**Comment by Nex-Pro** [comment]
Waiting for fresh build

**Comment by Nex-Pro** [comment]
@cain-kilgore Let's see if the fresh build fixes the issue.

**Comment by cain-kilgore** [comment]
@Nex-Pro: Can you try .setDominantHand("left") and ("right")?

**Comment by Nex-Pro** [comment]
@cain-kilgore There's no dominant hand setting in Avatar settings.

**Comment by cain-kilgore** [comment]
Oh that's weird.. Gustavo must have messed up. Yeah wait for a fresh build :)

**Comment by Nex-Pro** [comment]
[UncaughtException evaluate] Error: Result of expression 'MyAvatar.setDominantHand' [undefined] is not a function. in about:console:1 [Backtrace] <global>() at about:console:1

**Comment by cain-kilgore** [comment]
After this build now ^ should be good to go.

**Comment by Nex-Pro** [comment]
Hmm blame Gustavo ? He never even gets a thankyou :)

**Comment by Nex-Pro** [comment]
### FAIL

**Comment by cain-kilgore** [comment]
@Nex-Pro Should be good now. Working all good here on my side.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1888/HighFidelity-Beta-PR11029-9ff7891c880e94f593eba9db9d468c68fe188928.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1888/HighFidelity-Beta-PR11029-9ff7891c880e94f593eba9db9d468c68fe188928.dmg

**Comment by Nex-Pro** [comment]
Launch the Interface
Open the Console and Type MyAvatar.getDominantHand()
Confirm that it returns 'right'. **SUCCESS**
Go to Settings > Avatar
Confirm you see a Left and Right Hand Dominant Hand setting. **SUCCESS**
Confirm that one Radio Box must be ticked at all times, and unticking the already ticked Radio Box forces the selection to the Right Hand. **Do you mean left hand**
Confirm that when you select Left and Save Settings, that the Command From Step 2 returns 'left'.. **SUCCESS**
Open the Console and Type MyAvatar.setDominantHand("right")
Confirm that it sets successfully, and re-running the Command from Step 2 returns 'right' . **SUCCESS**
Check to ensure that each setting is updating in the Avatar Settings.. **SUCCESS**
Relaunch the game and confirm that the settings for both left and right stay put.. **SUCCESS**

For some reason this was reluctant to start on several runs. Had to repeat everything a couple of times to get a run-thru.

There is an issue with the dominant hand radio button behaviour.

Repeated clicks on the right hand keep that one active.
Repeated clicks on the left hand radio button toggles between right and left.

@cain-kilgore

**Comment by cain-kilgore** [comment]
Thanks @Nex-Pro:

The repeated clicks on the left hand radio button is expected; when you untick it it'll default to the Right Hand in order to ensure that one is always kept ticked.

**Comment by Nex-Pro** [comment]
@cain-kilgore Behaviour should be consistent surely ?

**Comment by Nex-Pro** [comment]
@cain-kilgore  Ask design.

**Comment by cain-kilgore** [comment]
Just about to do that now - sec!

**Comment by cain-kilgore** [comment]
@Nex-Pro: Design agrees by keeping the choice on the left. This next build now should have that.

**Comment by Nex-Pro** [comment]
@cain-kilgore  That;s why I'm here ! :)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1893/HighFidelity-Beta-PR11029-9c152ddc6f7d0c86c88b8423a8ee64f295351a29.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1893/HighFidelity-Beta-PR11029-9c152ddc6f7d0c86c88b8423a8ee64f295351a29.dmg

**Comment by cain-kilgore** [comment]
@Nex-Pro: Woo! Can you quickly test this PR now and confirm it's working the way you suggested now with the boxes?

**Comment by Nex-Pro** [comment]
Apologies for delays caused by my time getting eaten up with cmake/VS2017/qt5.9.1
Retesting right now.

**Comment by Nex-Pro** [comment]
As soon as this build is done.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1897/HighFidelity-Beta-PR11029-84609f27e36e161d4fb36039823396ec659e5fdf.exe

**Comment by cain-kilgore** [comment]
test this please

**Comment by Nex-Pro** [comment]
@cain-kilgore  failing this for now.  I think we should have these radio buttons in Toolbar mode and Tablet.
Currently they are absent in tablet.
Let me know if I'm incorrect.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1902/HighFidelity-Beta-PR11029-84609f27e36e161d4fb36039823396ec659e5fdf.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1902/HighFidelity-Beta-PR11029-84609f27e36e161d4fb36039823396ec659e5fdf.exe

**Comment by cain-kilgore** [comment]
@Nex-Pro New build sent out should fix these issues and correct the styling.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1909/HighFidelity-Beta-PR11029-280a7a7e42e5dabf8c6b284bcce4595c660ccd5d.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1909/HighFidelity-Beta-PR11029-280a7a7e42e5dabf8c6b284bcce4595c660ccd5d.dmg

**Comment by skytoad** [comment]
testing

**Comment by skytoad** [comment]
Test success!!

**Comment by cain-kilgore** [comment]
Yay!

**Comment by Nex-Pro** [comment]
@cain-kilgore This needs new testing notes

**Comment by Nex-Pro** [comment]
IMO

**Comment by cain-kilgore** [comment]
@Nex-Pro: I added one step to it if you'd like to recheck.

**Comment by skytoad** [comment]
Please note: I only tested the first 11 steps of this with the toolbar and it most likely needs tested on tablet as well

**Comment by skytoad** [comment]
Re-tested to include tablet: **Fail**

Dominant hand UI elements do not appear in tablet Settings > Avatar

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1915/HighFidelity-Beta-PR11029-2f857a6d1a8b58332eabc4e4329e30753abd17a3.dmg

**Comment by cain-kilgore** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1917/HighFidelity-Beta-PR11029-2f857a6d1a8b58332eabc4e4329e30753abd17a3.dmg

**Comment by cain-kilgore** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1918/HighFidelity-Beta-PR11029-2f857a6d1a8b58332eabc4e4329e30753abd17a3.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1918/HighFidelity-Beta-PR11029-2f857a6d1a8b58332eabc4e4329e30753abd17a3.dmg

**Comment by skytoad** [comment]
Testing

**Comment by skytoad** [comment]
**Fail:** Dominant hand UI elements still do not appear in tablet Settings > Avatar

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1921/HighFidelity-Beta-PR11029-f9204a7034d06a25391a74caf79cfb4ab272ab62.dmg

**Comment by cain-kilgore** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1922/HighFidelity-Beta-PR11029-f9204a7034d06a25391a74caf79cfb4ab272ab62.dmg

**Comment by cain-kilgore** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1923/HighFidelity-Beta-PR11029-f9204a7034d06a25391a74caf79cfb4ab272ab62.dmg

**Comment by cain-kilgore** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1925/HighFidelity-Beta-PR11029-f9204a7034d06a25391a74caf79cfb4ab272ab62.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1925/HighFidelity-Beta-PR11029-f9204a7034d06a25391a74caf79cfb4ab272ab62.exe

**Comment by Nex-Pro** [comment]
Retesting this

**Comment by Nex-Pro** [comment]
The testing notes of this PR remain incomplete.

However this is a **PASS**

In addition to the new radio buttons being present in tablet and toolbar modes, in both HMD and desktop, the radio buttons now behave correctly and meet the updated design criteria. IE they are now HiFi blue when active.

All gets via the console reflect the state of the buttons in the tablet.
All sets via the console are reflected in the Tablet and hud appropriately.
Dominant hand setting persists across restarts.

***Unable to locate "The game" in step 15 :)*** Substituted Interface

**Comment by cain-kilgore** [comment]
THE INTERFACE! Hahahaha, great testing. Thanks @Nex-Pro :D

**Pull request merged**